streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -252,6 +288,111 @@ public void buildAndOptimizeTopology() {
+                                                               parentSourceNode.valueSerde());
+
+            // re-use parent id to make sure the single repartition graph node is evaluated before downstream nodes
+            optimizedSingleRepartition.setId(parentSourceNode.id());
+
+            for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : streamsGraphNodeSetEntry.getValue()) {
+
+                final StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNodes().contains(keyChangingNode));
+
+                LOG.debug("Found the node downstream of key-changer {}", keyChangingNodeChild);
+
+                // need to add children of key-changing node as children of optimized repartition
+                // in order to process records from re-partitioning
+                optimizedSingleRepartition.addChildNode(keyChangingNodeChild);
+
+                LOG.debug("Removing {} from {}  children {}", keyChangingNodeChild, keyChangingNode, keyChangingNode.children());
+                // now remove children from key-changing node
+                keyChangingNode.removeChild(keyChangingNodeChild);
+
+                // now need to get children of repartition node so we can remove repartition node
+                final Collection<StreamsGraphNode> repartitionNodeToBeReplacedChildren = repartitionNodeToBeReplaced.children();
+
+                final StreamsGraphNode parentOfRepartitionNodeToBeReplaced = findParentNodeMatching(repartitionNodeToBeReplaced, n -> repartitionNodeToBeReplaced.parentNodes().contains(n));
+
+                if (parentOfRepartitionNodeToBeReplaced == null) {
+                    throw new IllegalStateException(String.format("Can't find parent node for %s this should not happen and indicates a bug", keyChangingNode));
+                }
+
+                LOG.debug("Returned {} parentOfRepartitionNodeToBeReplaced as parent of {}", parentOfRepartitionNodeToBeReplaced, repartitionNodeToBeReplaced);
+
+                // remove repartition node
+                parentOfRepartitionNodeToBeReplaced.removeChild(repartitionNodeToBeReplaced);
+
+                repartitionNodeToBeReplaced.clearChildren();
+
+                // re-attach children of removed repartition
+                for (final StreamsGraphNode repartitionNodeToBeReplacedChild : repartitionNodeToBeReplacedChildren) {
+                    parentOfRepartitionNodeToBeReplaced.addChildNode(repartitionNodeToBeReplacedChild);
+                }
+
+                LOG.debug("Updated node {} children {}", optimizedSingleRepartition, optimizedSingleRepartition.children());
+            }
+
+            keyChangingNode.addChildNode(optimizedSingleRepartition);
+            keyChangingOperationsToOptimizableRepartitionNodes.remove(streamsGraphNodeSetEntry.getKey());
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private StreamsGraphNode createRepartitionNode(final String name,
```

Comment:
Can we make this method return `OptimizableRepartitionNode` just for symmetry?

I.e., you can see that we're replacing a bunch of `OptimizableRepartitionNode`s with a new `OptimizableRepartitionNode` instead of replacing them with a generic `StreamsGraphNode`.

Reply:
ack